src/lwpa_plat/windows/lwpa/socket.c

```diff
@@ -669,45 +881,82 @@ int lwpa_poll(LwpaPollfd *fds, size_t nfds, int timeout_ms)
   {
     return kLwpaErrTimedOut;
   }
-  else if (fds && nfds > 0)
+  else
   {
-    for (fd = fds; fd < fds + nfds; ++fd)
+    lwpa_error_t res = kLwpaErrSys;
+    if (context->valid && lwpa_mutex_take(&context->lock, LWPA_WAIT_FOREVER))
     {
-      fd->revents = 0;
+      res = handle_select_result(context, event, &readfds, &writefds, &exceptfds);
+      lwpa_mutex_give(&context->lock);
+    }
+    return res;
+  }
+}
+
+lwpa_error_t handle_select_result(LwpaPollContext *context, LwpaPollEvent *event, const LwpaPollFdSet *readfds,
+                                  const LwpaPollFdSet *writefds, const LwpaPollFdSet *exceptfds)
+{
+  // Init the event data.
+  event->socket = LWPA_SOCKET_INVALID;
+  event->events = 0;
+  event->err = kLwpaErrOk;
+
+  // The default return; if we don't find any sockets set that we passed to select(), something has
+  // gone wrong.
+  lwpa_error_t res = kLwpaErrSys;
 
-      if (FD_ISSET(fd->fd, &readfds) || FD_ISSET(fd->fd, &writefds) || FD_ISSET(fd->fd, &exceptfds))
+  for (LwpaPollCtxSocket *sock_desc = context->sockets; sock_desc < context->sockets + context->socket_arr_size;
+       ++sock_desc)
+  {
+    if (sock_desc->socket == LWPA_SOCKET_INVALID)
+      continue;
+
+    if (LWPA_FD_ISSET(sock_desc->socket, readfds) || LWPA_FD_ISSET(sock_desc->socket, writefds) ||
+        LWPA_FD_ISSET(sock_desc->socket, exceptfds))
+    {
+      res = kLwpaErrOk;
+      event->socket = sock_desc->socket;
+      event->user_data = sock_desc->user_data;
+
+      /* Check for errors */
+      int error;
+      int error_size = sizeof(error);
+      if (getsockopt(sock_desc->socket, SOL_SOCKET, SO_ERROR, (char *)&error, &error_size) == 0)
       {
-        /* Check for errors */
-        int error;
-        int error_size = sizeof(error);
-        if (getsockopt(fd->fd, SOL_SOCKET, SO_ERROR, (char *)&error, &error_size) == 0)
+        if (error != 0)
         {
-          if (error != 0)
-          {
-            fd->revents |= LWPA_POLLERR;
-            fd->err = err_plat_to_lwpa(error);
-          }
-        }
-        else
-        {
-          return (int)err_plat_to_lwpa(WSAGetLastError());
+          event->events |= LWPA_POLL_ERR;
+          event->err = err_plat_to_lwpa(error);
         }
       }
-      if (nreadfds && (fd->events & LWPA_POLLIN) && FD_ISSET(fd->fd, &readfds))
+      else
+      {
+        res = err_plat_to_lwpa(WSAGetLastError());
+        break;
+      }
+      if (LWPA_FD_ISSET(sock_desc->socket, readfds))
       {
-        fd->revents |= LWPA_POLLIN;
+        if (sock_desc->events & LWPA_POLL_IN)
+          event->events |= LWPA_POLL_IN;
       }
-      if (nwritefds && (fd->events & LWPA_POLLOUT) && FD_ISSET(fd->fd, &writefds))
+      if (LWPA_FD_ISSET(sock_desc->socket, writefds))
       {
-        fd->revents |= LWPA_POLLOUT;
+        if (sock_desc->events & LWPA_POLL_CONNECT)
```

Comment:
I'm a bit confused by LWPA_POLL_CONNECT.  When does that event get cleared?  Otherwise the user will never see LWPA_POLL_OUT, right?

Reply:
LWPA_POLL_CONNECT is an abstraction around the fact that different operating systems require different polling methods for nonblocking connects. For example, on Windows using select(), you need to add a connecting socket to both writefds and exceptfds to handle errors. On other platforms I believe it's only writefds.

So this flag is necessary to differentiate nonblocking connects from general polling for writability (i.e. a write to a full TCP pipe returns EWOULDBLOCK, and the app wants to poll for when there is buffer space again).

Reply:
More to your point, theoretically once the socket is connected, the user will modify the socket so that it is not interested in the LWPA_POLL_CONNECT event anymore. Then they can utilize LWPA_POLL_OUT if desired.

Reply:
I'd be open to the idea of making LWPA_POLL_CONNECT an auto-clearing event...

Reply:
How do the modify the socket?  Do they have to remove & re-add it?

Reply:
> How do the modify the socket? Do they have to remove & re-add it?

Right now yes, I believe that's how they'd have to do it.

Reply:
It feels cleaner to me to have a function to clear the CONNECT flag.  I could be persuaded to make this auto-clearing, too, since you presumably want to write once you've connected..

Reply:
As far as the function to clear goes, how about just a generic modify_socket() function where you can change the set of events however you want?

Reply:
I originally wasn't sure if you wanted to add such a general feature for the specific CONNECT issue, but that I think that's probably the right way to do it.